calyx/src/passes/infer_share.rs

```diff
@@ -0,0 +1,386 @@
+use crate::analysis::{DominatorMap, ReadWriteSet, ShareSet};
+use crate::errors::CalyxResult;
+use crate::ir;
+use crate::ir::traversal::{
+    Action, ConstructVisitor, Named, VisResult, Visitor,
+};
+use crate::ir::CloneName;
+use std::collections::HashSet;
+use std::rc::Rc;
+
+const BEGIN_ID: &str = "BEGIN_ID";
+const END_ID: &str = "END_ID";
+
+// Inputs are a control statement c and a u64 id. If control is an if statment, then
+// the id should refer to either the begin or end id of stmt c. Returns true if id refers
+// to the begin id and false if it refers to the end id. If it is not an if statement, behavior
+// is unspecified.
+fn is_begin_id(c: &ir::Control, id: u64) -> bool {
+    match c {
+        ir::Control::If(if_control) => {
+            if let Some(&begin) = if_control.attributes.get(BEGIN_ID) {
+                if begin == id {
+                    return true;
+                }
+            }
+            if let Some(&end) = if_control.attributes.get(END_ID) {
+                if end == id {
+                    return false;
+                }
+            }
+            unreachable!("id should match either beginning or ending id")
+        }
+        _ => true,
+    }
+}
+
+pub struct InferShare {
```

Comment:
This needs a documentation comment (`///`) with a short explanation of what the pass is doing. It's okay to just refer to the DominatorMap logic while explaining it.